src/naming/Naming_AST.ml

```diff
@@ -840,7 +840,8 @@ let resolve lang prog =
                       let sid = SId.unsafe_default in
                       let rest_of_middle = Common.map fst rest_of_middle in
                       let canonical =
-                        xs @ dotted_to_canonical (rest_of_middle @ [ id ])
+                        canonical_append xs
```

Comment:
So here given `canonical_append`s definition we throw away `xs`'s token, and keep  the one from `dotted_to_canonical ...` which is guaranteed to be `None`. Is that fine?

Reply:
It's meant to admit easy refactoring even if it is not. Should it end up being that we need that token around, we should just change `dotted_to_canonical` to embed the final token into the `canonical_name`.

I believe that this is not actually necessary here, since we are erring on the side of not allowing unique characteristics for these canonical names, in the open-source engine. These canonical names may pass back unique characteristics from the deep engine, but here they aren't so important.